Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1081 @@
+           _Py_UOpsAbstractFrame *frame,
+           _Py_UOpsSymType **localsplus_start,
+           int locals_len,
+           int curr_stacklen,
+           int total_len)
+{
+    frame->locals = localsplus_start;
+    frame->stack = frame->locals + locals_len;
+    frame->stack_pointer = frame->stack + curr_stacklen;
+    ctx->water_level = localsplus_start + total_len;
+    if (ctx->water_level > ctx->limit) {
+        return -1;
+    }
+    return 0;
+}
+
+static inline int
+frame_initalize(_Py_UOpsAbstractInterpContext *ctx, _Py_UOpsAbstractFrame *frame,
+                int locals_len, int curr_stacklen)
+{
+    // Initialize with the initial state of all local variables
+    for (int i = 0; i < locals_len; i++) {
+        _Py_UOpsSymType *local = sym_init_unknown(ctx);
+        if (local == NULL) {
+            goto error;
+        }
+        frame->locals[i] = local;
+    }
+
+
+    // Initialize the stack as well
+    for (int i = 0; i < curr_stacklen; i++) {
+        _Py_UOpsSymType *stackvar = sym_init_unknown(ctx);
+        if (stackvar == NULL) {
+            goto error;
+        }
+        frame->stack[i] = stackvar;
+    }
+
+    return 0;
+
+error:
+    return -1;
+}
+
+static inline _Py_UOpsAbstractFrame *
+frame_new(_Py_UOpsAbstractInterpContext *ctx,
+                          PyObject *co_consts, int stack_len, int locals_len,
+                          int curr_stacklen)
+{
+    _Py_UOpsSymType **sym_consts = create_sym_consts(ctx, co_consts);
+    if (sym_consts == NULL) {
+        return NULL;
+    }
+    _Py_UOpsAbstractFrame *frame = PyObject_New(_Py_UOpsAbstractFrame,
+                                                      &_Py_UOpsAbstractFrame_Type);
+    if (frame == NULL) {
+        PyMem_Free(sym_consts);
+        return NULL;
+    }
+
+
+    frame->sym_consts = sym_consts;
+    frame->sym_consts_len = (int)Py_SIZE(co_consts);
+    frame->stack_len = stack_len;
+    frame->locals_len = locals_len;
+    frame->prev = NULL;
+    frame->next = NULL;
+
+    return frame;
+}
+
+static inline bool
+sym_is_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ);
+static inline uint64_t
+sym_type_get_refinement(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ);
+
+static inline PyFunctionObject *
+extract_func_from_sym(_Py_UOpsSymType *callable_sym)
+{
+    assert(callable_sym != NULL);
+    if (!sym_is_type(callable_sym, PYFUNCTION_TYPE_VERSION_TYPE)) {
+        DPRINTF(1, "error: _PUSH_FRAME not function type\n");
+        return NULL;
+    }
+    uint64_t func_version = sym_type_get_refinement(callable_sym, PYFUNCTION_TYPE_VERSION_TYPE);
+    PyFunctionObject *func = _PyFunction_LookupByVersion((uint32_t)func_version);
+    if (func == NULL) {
+        OPT_STAT_INC(optimizer_failure_reason_null_function);
+        DPRINTF(1, "error: _PUSH_FRAME cannot find func version\n");
+        return NULL;
+    }
+    return func;
+}
+
+
+// 0 on success, anything else is error.
+static int
+ctx_frame_push(
```

Comment:
I still don't like how this works, sorry. In `translate_bytecode_to_trace()` I use a fixed array of 5 frame structures (`trace_stack`). You could probably do that too here, and it would no longer have to be an object in a doubly-linked list. (I think the frame depth you will see here is limited by what the translator does, so you should be able to use the same number of frames in the array.

(Also the code layout here is a bit hard to read.)